[Mo Yunbin] iP

yunbinmo opened the pull request:
# Spike
> "Don't be pushed around by the fears in your mind. Be led by the dreams in your heart.”― Roy T. Bennett

The chatbot is:
- Text-based
- Quick to learn

Here's how to use it:
1. Download it from [here](https://github.com/yunbinmo/ip)
2. Double click on the `.jar` file
3. Let's Go!👍

It is **FREE** and more *features* coming soon

Spike can help you with the following things:
- [x]  Create a task list
- [x] Manage the tasks
- [x] Save your tasks


If your are a developer or Java learner, a quick review of the main method:
```Java
public static void main(String[] args) {
    new Spike("data/", "data/Spike.txt").run();
    return;
}
```

Commit:
Add Gradle support

yunbinmo committed:
Implement level 1 - Greet, Echo, Exit

yunbinmo committed:
Implement level 2 - Add, List

yunbinmo committed:
Implement level-3 - Mark as Done/Undone, add class Task

yunbinmo committed:
Implement level 4 - Todo, Event, Deadline with inheritance

yunbinmo committed:
Update automated test files and fix minor bugs in previous commit

yunbinmo committed:
Implement level 5 - handle errors, add DukeException

yunbinmo committed:
Implement level 6 - delete

yunbinmo committed:
Fix a minor bug in the previous level-6 commit

yunbinmo committed:
Add enum to centralise the processing of commands

yunbinmo committed:
Fix some coding segment to comply with coding standards

yunbinmo committed:
Implement level-7 Save

yunbinmo committed:
Implement level-8 Dates and Times

yunbinmo committed:
Change Git ignore list

yunbinmo committed:
Merge branch 'branch-Level-7'

yunbinmo committed:
Merge branch 'branch-Level-8'

yunbinmo committed:
Refactor the code and divide class into packages

yunbinmo committed:
Add JUnit Tests for Ui and Parser class

yunbinmo committed:
Add JavaDoc for public methods

yunbinmo committed:
Check compliance with coding standard

yunbinmo committed:
Change Git ignore list

yunbinmo committed:
Implement leve-9 find

yunbinmo committed:
Merge branch 'branch-A-JavaDoc'

yunbinmo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/spike/command/Command.java

yunbinmo committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/spike/command/AddCommand.java
#	src/main/java/spike/command/DeleteCommand.java
#	src/main/java/spike/command/ListCommand.java

yunbinmo committed:
Add more JUnit tests

yunbinmo committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

yunbinmo committed:
Add checkstyle files

yunbinmo committed:
Merge branch 'branch-A-CheckStyle'

yunbinmo committed:
Update .gitignore file

yunbinmo committed:
Update build.gradle

yunbinmo committed:
Reconcile changes for restarting the branch-A-Gradle

# Conflicts:
#	build.gradle

yunbinmo committed:
Update .gitignore file

yunbinmo committed:
Integrate Gradle with the project

sibinhho99-nus commented on src/main/java/spike/command/FindCommand.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * Constructor using keyword string.
+     *
+     * @param keyWord the word used to find tasks
```
`keyword` should be one word, so `W` should not be capitalized

yunbinmo replied:
Thanks for the comments, will fix other issues too.

sibinhho99-nus commented on src/main/java/spike/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    /**
+     * Format the string representation of spike.task.Task objects.
```
Is this a stray line?

sibinhho99-nus commented on src/test/java/spike/ui/UiTest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Test
+    public void getCommand_emptyCommand_getNextLine() {
```
You should perhaps add header comments here

yunbinmo committed:
Implement Level-10: GUI

yunbinmo committed:
Merge branch 'branch-Level-10'

dannytayjy commented on src/main/java/spike/Spike.java:
```diff
@@ -0,0 +1,81 @@
+
+    /**
+     * Gets number of tasks.
+     * @return
```
Incomplete Javadoc comments

yunbinmo replied:
Thanks for the comments, will fix them later!

dannytayjy commented on src/main/java/spike/command/AddCommand.java:
```diff
@@ -0,0 +1,38 @@
+    public String execute(TaskList tasks) {
+        tasks.addTask(task);
+        return getAddedTaskText(task, tasks);
+    }
```
Missing Javadoc comments

dannytayjy commented on src/main/java/spike/command/FindCommand.java:
```diff
@@ -0,0 +1,39 @@
+     * @return matching tasks in string
+     */
+    @Override
+    public String execute(TaskList tasks) {
+        int i = 1;
+        String result = "Here are the matching tasks in your list:\n";
+        for (Task task : tasks.getTasks()) {
+            if (task.toString().contains(keyWord)) {
+                result = result + i + "." + task + "\n";
+            }
+            i++;
+        }
+        return result.trim();
+    }
```
You might want to put some line breaks to improve readability

dannytayjy replied:
I see this problem at other places too. Since the code is relatively short, it might be fine, but I suggest putting a line break before and after the for loop.

dannytayjy commented on src/main/java/spike/command/ListCommand.java:
```diff
@@ -0,0 +1,76 @@
+     * Lists all tasks.
+     *
+     * @param tasks
+     * @return
```
Incomplete Javadoc comments

dannytayjy commented on src/main/java/spike/command/ListCommand.java:
```diff
@@ -0,0 +1,76 @@
+        return result.trim();
+    }
+
+    private String getTaskListTextByDate(TaskList tasks) {
+        int i = 1;
+        String result = "Here are the task(s) in your list filtered by date:\n";
+        for (Task task : tasks.getTasks()) {
+            if (task.getDateTime().toLocalDate().equals(ldt.toLocalDate())) {
+                result = result + i + "." + task + "\n";
+            }
+            i++;
+        }
+        return result.trim();
+    }
```
Missing Javadoc comments

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+package spike.parser;
+import java.time.LocalDateTime;
```
I suggest putting a line break here

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+            try {
+                return parseList(inputLine);
+            } catch (SpikeException e) {
+                return new IncorrectCommand(e.getMessage());
+            }
+        case MARK:
+            // Fallthrough
+        case UNMARK:
+            try {
+                return parseToggleMark(type, commandWords, tasks);
+            } catch (SpikeException e) {
+                return new IncorrectCommand(e.getMessage());
+            }
+        case DELETE:
+            try {
+                return parseDelete(commandWords, tasks);
+            } catch (SpikeException e) {
+                return new IncorrectCommand(e.getMessage());
+            }
+        case TODO:
+            // Fallthrough
+        case DEADLINE:
+        case EVENT:
+            try {
+                return parseAdd(type, inputLine, commandWords);
+            } catch (SpikeException e) {
+                return new IncorrectCommand(e.getMessage());
+            }
+        case FIND:
+            try {
+                return parseFind(inputLine);
+            } catch (SpikeException e) {
+                return new IncorrectCommand(e.getMessage());
+            }
+        case BYE:
+            return parseExit();
+        default:
+            return new IncorrectCommand("Should not reach this line");
+        }
+    }
```
You might want to put some line breaks to improve readability

dannytayjy commented on src/main/java/spike/storage/Storage.java:
```diff
@@ -0,0 +1,118 @@
+import spike.task.TaskList;
+import spike.task.ToDo;
+
+
+
```
You can remove these additional line breaks

dannytayjy commented on src/main/java/spike/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+        this.dateTime = dateTime;
+    }
+
+    /**
+     * Retrieve the task status icon.
+     */
+    private String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
```
Incomplete Javadoc comments

dannytayjy commented on src/main/java/spike/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+        this.isDone = false;
+    }
+
+    /**
+     * Returns the representative string for saving in data file.
+     */
+    public String toFileFormat() {
+        return "";
+    }
```
Incomplete Javadoc comments

dannytayjy commented on src/main/java/spike/task/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+        this.tasks.remove(task);
+    }
+
+    /**
+     * Returns all task in the form of arraylist
+     */
+    public ArrayList<Task> getTasks() {
+        return this.tasks;
+    }
```
Incomplete Javadoc comments

dannytayjy commented on src/main/java/spike/task/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+        return this.tasks;
+    }
+
+
```
You can remove the additional line break

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    private Command parseFind(String inputLine) throws SpikeException {
+        if (inputLine.length() <= 5) {
+            throw new SpikeException("Kindly enter the keyword for finding task");
+        }
+        return new FindCommand(inputLine.substring(5));
+    }
```
Missing Javadocs comments

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+
+    private Command parseList(String inputLine) throws SpikeException {
+        if (inputLine.length() >= 5) {
+            // User tries to list task by date
+            DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+            LocalDateTime ldt = parseDateTime(inputLine.substring(5), dtf);
+            if (ldt == null) {
+                throw new SpikeException("Kindly enter the date in the format yyyy-MM-dd 0000 to filter by date");
+            }
+            return new ListCommand(0, ldt);
+        } else {
+            return new ListCommand();
+        }
+    }
```
Missing Javadocs comments

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+    private Command parseAdd(CommandName c, String command, String[] commandWords) throws SpikeException {
+        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+        switch (c) {
+        case TODO:
+            if (command.length() <= 5) {
+                throw new SpikeException("Hmmmm what to do? Think again?");
+            }
+            ToDo newTD = new ToDo(command.substring(command.indexOf("todo") + 5));
+            return new AddCommand(newTD);
+        case DEADLINE:
+            // Extract description and deadline and pass to constructor
+            if (commandWords.length <= 2 || command.indexOf("/by") == -1
+                    || commandWords[1].equals("/by") || command.indexOf("/by") + 3 == command.length()) {
+                throw new SpikeException("Deadline or task description missing.");
+            }
+            LocalDateTime deadlineT = parseDateTime(command.substring(command.indexOf("/by") + 4), dtf);
+            if (!(deadlineT == null)) {
+                Deadline newD = new Deadline(command.substring(command.indexOf("deadline") + 9,
+                        command.indexOf("/by") - 1), deadlineT);
+                return new AddCommand(newD);
+            } else {
+                throw new SpikeException("Please enter a valid date in the format yyyy-MM-dd HHmm");
+            }
+        case EVENT:
+            if (commandWords.length <= 2 || command.indexOf("/at") == -1
+                    || commandWords[1].equals("/at") || command.indexOf("/at") + 3 == command.length()) {
+                throw new SpikeException("Event time or event description missing.");
+            }
+            LocalDateTime eventT = parseDateTime(command.substring(command.indexOf("/at") + 4), dtf);
+            if (!(eventT == null)) {
+                Event newE = new Event(command.substring(command.indexOf("event") + 6,
+                        command.indexOf("/at") - 1), eventT);
+                return new AddCommand(newE);
+            } else {
+                throw new SpikeException("Please enter a valid date in the format yyyy-MM-dd HHmm");
+            }
+        default:
+            return new IncorrectCommand("Should not reach this line");
+        }
+    }
```
- Missing Javadocs comments
- You might want to put some line breaks to improve readability

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    private Command parseDelete(String[] commandWords, TaskList tasks) throws SpikeException {
+        if (commandWords.length != 2 || isInt(commandWords[1]) == -1
+                || isInt(commandWords[1]) > tasks.getListSize()) {
+            throw new SpikeException("Invalid arguments for deletion. Please check again!");
+        }
+        Task toDelete = tasks.getTasks().get(Integer.parseInt(commandWords[1]) - 1);
+        return new DeleteCommand(toDelete);
+    }
```
Missing Javadocs comments

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+        return new DeleteCommand(toDelete);
+    }
+
+    private Command parseToggleMark(CommandName c, String[] commandWords, TaskList tasks) throws SpikeException {
+        if (c.getCommand().equals("mark")) {
+            if (commandWords.length != 2 || isInt(commandWords[1]) == -1
+                    || isInt(commandWords[1]) > tasks.getListSize()) {
+                throw new SpikeException("Invalid arguments for marking. Please check again!");
+            }
+            Task toMark = tasks.getTasks().get(Integer.parseInt(commandWords[1]) - 1);
+            return new ToggleMarkCommand(1, toMark);
+        } else {
+            if (commandWords.length != 2 || isInt(commandWords[1]) == -1
+                    || isInt(commandWords[1]) > tasks.getListSize()) {
+                throw new SpikeException("Invalid arguments for unmarking. Please check again!");
+            }
+            Task toUnmark = tasks.getTasks().get(Integer.parseInt(commandWords[1]) - 1);
+            return new ToggleMarkCommand(0, toUnmark);
+        }
+    }
```
Missing Javadocs comments

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    private Command parseExit() {
+        return new ExitCommand();
+    }
```
Missing Javadocs comments

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+
+    /**
+     * Parses date and time input by user and returns valid LocalDateTime object
+     */
+    private static LocalDateTime parseDateTime(String s, DateTimeFormatter dtf) {
+        try {
+            LocalDateTime dateTime = LocalDateTime.parse(s, dtf);
+            return dateTime;
+        } catch (DateTimeParseException e) {
+            return null;
+        }
+    }
```
Incomplete Javadocs comments, `@return` is missing (same as others which are incomplete)

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    /**
+     * Checks whether the input string is integer.
+     * If yes, return it, else return -1.
+     */
+    private static int isInt(String str) {
+        try {
+            int x = Integer.parseInt(str);
+            return x; // it is an integer
+        } catch (NumberFormatException e) {
+            return -1; // not an integer
+        }
+
+    }
```
Incomplete Javadocs comments, `@return` is missing (same as others which are incomplete)

dannytayjy commented on src/main/java/spike/parser/Parser.java:
```diff
@@ -0,0 +1,238 @@
+
+    }
+
+    /**
+     * Checks whether it is a valid command.
+     * If valid, return that command enum number, else return null.
+     */
+    private CommandName validateCommand(String input) {
+        for (CommandName c : CommandName.values()) {
+            if (c.getCommand().equals(input)) {
+                return c;
+            }
+        }
+        return null;
+    }
```
Incomplete Javadocs comments, `@return` is missing (same as others which are incomplete)

dannytayjy commented on src/main/java/spike/command/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    /**
+     * Removes and printed the deleted task.
```
"printed" should be "prints"

dannytayjy commented on src/main/java/spike/command/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+        return getDeletedTaskText(task, tasks);
+    }
+
+    /**
+     * Returns the message needed for printing when deleting task.
+     */
```
Incomplete Javadoc comments

dannytayjy commented on src/main/java/spike/command/ExitCommand.java:
```diff
@@ -0,0 +1,20 @@
+ * Handles data before exiting.
+ */
+public class ExitCommand extends Command {
+    public static final String EXIT_MESSAGE = "See you soon! ﾍ(=￣∇￣)ﾉ You may close this window now.";
+    /**
```
You might want to put a line break to improve readability

yunbinmo replied:
Thanks again for all the detailed inputs.

dannytayjy reviewed (commented):
I think your codes are generally good. Not many issues with coding standards.

General issues:
- Incomplete or missing Javadocs comments
- Try and add some line breaks where appropriate to improve readability

seksek13 commented on src/main/java/Spike.java:
```diff
@@ -108,34 +168,51 @@ public static Command isValidCommand(String command) {
-        case "deadline":
+        case DEADLINE:
             // Extract description and deadline and pass to constructor
             if (commandWords.length <= 2 || command.indexOf("/by") == -1
```
is it possible to seperate the comparisons so that it would be easier for others to understand

yunbinmo replied:
Good idea, I will look into this.

seksek13 commented on src/main/java/Spike.java:
```diff
@@ -1,7 +1,14 @@
+import java.io.File;
```
I like that you had import the functions explicitly, it makes it easier to see which classes are used.

seksek13 commented on src/main/java/spike/Main.java:
```diff
@@ -0,0 +1,32 @@
+public class Main extends Application {
+
+    private Spike spike = new Spike("/data", "/data/Spike.txt");
+
```
It would be better to add some headers comment here :)

seksek13 commented on src/main/java/spike/command/DeleteCommand.java:
```diff
@@ -27,7 +27,7 @@ public DeleteCommand(Task task) {
+        tasks.deleteTask(task);
         return getDeletedTaskText(task, tasks);
     }
 
```
Methods and variables are clean and easy to understand

seksek13 reviewed (commented):
Overall, there is a good practice of correct coding standards, it will be great when the javadocs are added those methods mentioned, good job!

yunbinmo committed:
Fix some coding standard and javadoc issues

yunbinmo committed:
Add assertions at some points of the code base

yunbinmo committed:
Refactor to improve code quality

yunbinmo committed:
Use Java Stream to replace some loop operations

yunbinmo committed:
Merge pull request #1 from yunbinmo/branch-A-Assertions

Add assertions at some points of the code base

yunbinmo committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/spike/parser/Parser.java

yunbinmo committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/spike/command/FindCommand.java

yunbinmo committed:
Merge pull request #3 from yunbinmo/branch-A-Streams

Use Java Stream to replace some loop operations

yunbinmo committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/spike/command/FindCommand.java
#	src/main/java/spike/command/ListCommand.java

yunbinmo committed:
Merge pull request #2 from yunbinmo/branch-A-CodeQuality

Refactor to improve code quality

yunbinmo committed:
Create gradle.yml

yunbinmo committed:
Update gradle.yml

yunbinmo committed:
Update gradle.yml

yunbinmo committed:
Update gradle.yml

yunbinmo committed:
Update gradle.yml

yunbinmo committed:
Update gradle.yml

yunbinmo committed:
Add simple reminder functionality

yunbinmo committed:
Merge branch 'branch-B-Reminders'

yunbinmo committed:
Merge branch 'master' of https://github.com/yunbinmo/ip

yunbinmo committed:
Parser class: extract constant variables

Methods in the Parser are using too many magic literals.

Let's extract the magic strings as constants in the Parser class.

Using constants make changes to the response message more convenient.

yunbinmo committed:
Add assertions to validate assumptions made

Some parts of the code are written based on assumptions made on
parameters passed into the method, an additional guard needs to be
added to make sure unexpected error could be detected as early as
possible.

Let's add assertions at the beginning of some methods where necessary.

Adding assertions will avoid unexpected error (if any) being invisible
during execution, and therefore prevent further errors from happening.

yunbinmo committed:
Replace for loops with Java Stream where possible

Stream provides more elegant ways of processing a sequence of items,
making it easier for any future modification.

Let's replace for loops with methods from the Stream class
(e.g. forEach(), filter(), and orElse())to keep the original behaviours.

The processing done on the task item now looks cleaner and make more
sense for potential future developers.

yunbinmo committed:
Update build.gradle file

yunbinmo committed:
Update build.gradle file

yunbinmo committed:
Update build.gradle

yunbinmo committed:
Implement GUI of Spike

yunbinmo committed:
Update checkstyle.xml

yunbinmo committed:
Fix some JavaDoc

yunbinmo committed:
Add files via upload

yunbinmo committed:
Delete config/checkstyle directory

yunbinmo committed:
Update gitignore file and docs

yunbinmo committed:
Merge branch 'master' of https://github.com/yunbinmo/ip

yunbinmo committed:
Update README.md

yunbinmo committed:
Update README.md

yunbinmo committed:
Update README.md

yunbinmo committed:
Update README.md

yunbinmo committed:
Fix some bugs for final release

yunbinmo committed:
Update README.md
